### ISSUE TITLE
Plugins without #[no_link] produce shared library error when running

### ISSUE THREAD

**Issue by daboross** [rust-lang/rust#20769]
When building a project using regex_macros, the binary runs successfully on my development system, but fails on production (where rust isn't installed).

Here's the smallest program I could think of that will cause this:
```rust
#![feature(plugin)]
extern crate regex;
#[plugin]
// works correctly if #[no_link] is used here
extern crate regex_macros;

fn main() {
}
```
An even more concise example causing the same effect:
```rust
extern crate regex_macros;
fn main() {}
```
That program compiles and runs successfully on a system with rust installed, but fails on one which isn't with the following error message:
```
$ ./test-project 
./test-project: error while loading shared libraries: libstd-4e7c5e5c.so: cannot open shared object file: No such file or directory
```

Note that this works correctly if `#[no_link]` is added before `extern crate regex_macros;`. This also works correctly when regex_macros isn't used.

I wasn't sure if this is an issue in rust or the regex library, so I've submitted the issue here. Feel free to close it and direct me to the correct place to submit this.

**Comment by steveklabnik** [comment]
Are you using regex_macros from crates.io? You're not supposed to use the in-tree one any more.

**Comment by sfackler** [comment]
This is (sort of) expected behavior. Most crates are compiled into rlibs and statically linked into the final result, while compiler plugins like regex_macros are dynamic libraries since they're loaded into the compiler at runtime.

The real issue here is that the directive to import a compiler plugin should not runtime link to that plugin by default, since it's almost never what you want. @kmcallister and I discussed this a bit in IRC and I believe agreed that the current setup with a required `#[no_link]` is a bummer, but decided to punt on thinking too much about the right way to handle this until after the 1.0 rush is over, since plugins are going to be feature gated for that release anyway.

**Comment by kmcallister** [comment]
Yeah, `#[no_link]` is a temporary measure while we think about whether we want plugins to be `extern crate`s at all.

**Comment by daboross** [comment]
@steveklabnik Yep, I'm using the crates.io one.

@sfackler OK. Do you know where I can file an issue about fixing the regex documentation to specify that #[no_link] should be used? It currently still shows just #[plugin].

**Comment by sfackler** [comment]
https://github.com/rust-lang/regex/issues

**Comment by kmcallister** [comment]
Will be fixed by #21043.
